infra/cifuzz/cifuzz.py

```diff
@@ -124,35 +171,48 @@ def build_fuzzers(  # pylint: disable=too-many-arguments,too-many-locals
   out_dir = os.path.join(workspace, 'out')
   os.makedirs(out_dir, exist_ok=True)
 
-  # Build Fuzzers using docker run.
-  inferred_url, project_builder_repo_path = (
-      build_specified_commit.detect_main_repo(project_name,
-                                              repo_name=project_repo_name))
-  if not inferred_url or not project_builder_repo_path:
-    logging.error('Could not detect repo from project %s.', project_name)
-    return False
-  project_repo_name = os.path.basename(project_builder_repo_path)
-  src_in_project_builder = os.path.dirname(project_builder_repo_path)
-
-  manual_src_path = os.getenv('MANUAL_SRC_PATH')
-  if manual_src_path:
-    if not os.path.exists(manual_src_path):
+  project_src_path = get_project_src_path()
+  if project_src_path:
+    if not os.path.exists(project_src_path):
       logging.error(
-          'MANUAL_SRC_PATH: %s does not exist. '
-          'Are you mounting it correctly?', manual_src_path)
+          'PROJECT_SRC_PATH: %s does not exist. '
+          'Are you mounting it correctly?', project_src_path)
       return False
-    # This is the path taken outside of GitHub actions.
-    git_workspace = os.path.dirname(manual_src_path)
+    git_workspace = os.path.dirname(project_src_path)
   else:
     git_workspace = os.path.join(workspace, 'storage')
     os.makedirs(git_workspace, exist_ok=True)
 
-  # Checkout projects repo in the shared volume.
+  build_integration_path = os.getenv('BUILD_INTEGRATION_PATH')
+  if build_integration_path:
+    build_integration_path = os.path.join(project_src_path,
+                                          build_integration_path)
+    logging.info('build_integration_path %s, project_src_path %s.',
+                 build_integration_path, project_src_path)
+    logging.info('Building external project.')
+    inferred_url = None
+    project_builder_repo_path = os.path.join('/src',
+                                             os.path.basename(project_src_path))
+    if not build_external_project_docker_image(project_name, project_src_path,
+                                               build_integration_path):
+      return False
+  else:
+    # detect_main_repo builds the image as a side effect.
+    inferred_url, project_builder_repo_path = (
+        build_specified_commit.detect_main_repo(project_name,
+                                                repo_name=project_repo_name))
+
+    if not inferred_url or not project_builder_repo_path:
+      logging.error('Could not detect repo from project %s.', project_name)
+      return False
+  project_repo_name = os.path.basename(project_builder_repo_path)
```

Comment:
Another suggestion: it's rather strange that we pass `project_repo_name` which ends up being completely ignored and replaced here. A single function shouldn't be dealing with all this complexity.

I would recommend refactoring this such such that all these complex logic is moved elsewhere, and the results are passed to this _do_build_fuzzers function, which should be kept simple. we can still have the `build_fuzzers` convenience function like so:

```python
def build_fuzzers(...):
  project_src_path = ...
  main_repo_url, src_in_project_builder = cifuzz.find_main_repo_url(...)
  _do_build_fuzzers(..., main_repo_url, src_in_project_builder, project_src_path...)
```

That way the logic is easier to follow and this probably becomes easier to test and extend as well. Would something like that be possible?

Reply:
+1

Reply:
I think this code was terrible because I was forcing this function to deal with three different cases while trying to minimize the amount of duplicated code. I've since replace that model [with one where there is a base class and a child for each of the three cases (OSS-Fuzz+Github, OSS-Fuzz+GenericCI, non-oss-fuzz+Github)](https://github.com/google/oss-fuzz/pull/4656/files#diff-149370d83b6b10e6a0c75ec5ac0cddd5456830259b9c35692327acb8e60f1497R133).